src/zhttp_server.c

```diff
@@ -0,0 +1,512 @@
+/*  =========================================================================
+    zhttp_server -
+
+    Copyright (c) the Contributors as noted in the AUTHORS file.
+
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+    =========================================================================
+*/
+
+/*
+@header
+    zhttp_server -
+@discuss
+@end
+*/
+
+#include "czmq_classes.h"
+
+#ifdef HAVE_LIBMICROHTTPD
+#include <microhttpd.h>
+#endif
+
+#ifdef HAVE_LIBMICROHTTPD
+
+typedef struct {
+    char* content;
+    size_t content_length;
+    zhash_t *headers;
+} request_t;
+
+typedef struct {
+    zsock_t *pipe;              //  Actor command pipe
+    bool terminated;            //  Did caller ask us to quit?
+    bool verbose;               //  Verbose logging enabled?
+
+    zhttp_server_options_t *options;
+    zsock_t *backend;
+    struct MHD_Daemon *daemon;
+    zframe_t *empty_frame;
+    zlistx_t* pending_connections;
+} server_actor_t;
+
+static int
+s_handle_request (
+        server_actor_t *self,
+        struct MHD_Connection *mhd_connection,
+        const char *url,
+        const char *method,
+        const char *version,
+        const char *upload_data,
+        size_t *upload_data_size,
+        request_t **request_p);
+
+static void
+s_handle_request_completed (
+        server_actor_t *self,
+        struct MHD_Connection *mhd_connection,
+        request_t **request_p,
+        enum MHD_RequestTerminationCode toe);
+
+static void
+server_actor (zsock_t *pipe, void *args);
+
+
+static request_t*
+request_new () {
+    request_t *self = (request_t*)zmalloc (sizeof (request_t));
+    assert (self);
+
+    self->headers = zhash_new ();
+    zhash_autofree (self->headers);
+
+    return self;
+}
+
+static void
+request_destroy (request_t **self_p) {
+    request_t * self= *self_p;
+
+    if (self) {
+        zhash_destroy (&self->headers);
+        zstr_free (&self->content);
+
+        free (self);
+    }
+
+    *self_p = NULL;
+}
+
+static void *
+s_insert_connection (struct MHD_Connection *connection) {
+    MHD_suspend_connection (connection);
+    return connection;
+}
+
+static void
+s_destroy_connection (struct MHD_Connection **connection) {
+    MHD_resume_connection (*connection);
+    *connection = NULL;
+}
+
+static server_actor_t *
+server_actor_new (zsock_t *pipe, zhttp_server_options_t *options)
+{
+    server_actor_t *self = (server_actor_t *) zmalloc (sizeof (server_actor_t));
+    assert (self);
+
+    self->daemon = MHD_start_daemon (
+            MHD_ALLOW_SUSPEND_RESUME, // TODO: use epoll on linux
```

Comment:
@somdoron this was introduced in 0.9.34 by https://github.com/scottjg/libmicrohttpd/commit/b19e9c0cf0a3aef3794b5994288b0478b20e9490 - and SUSE until Leap 15 has 0.9.33. Can this flag be omitted if not available, or will it break things?

Reply:
I think we just need to have another version flag, if less than 0.9.34 we should use MHD_USE_PIPE_FOR_SHUTDOWN instead.

Reply:
ok, can send a PR in a minute

Reply:
Cool